clippy_lints/src/assert_ok.rs

```diff
@@ -0,0 +1,81 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::macros::{find_assert_args, root_macro_call_first_node, PanicExpn};
+use clippy_utils::source::snippet_opt;
+use clippy_utils::ty::{implements_trait, is_type_diagnostic_item};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self, Ty};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `assert!(r.is_ok())` calls.
+    ///
+    /// ### Why is this bad?
+    /// An assertion failure cannot output a useful message of the error.
+    ///
+    /// ### Example
+    /// ```rust,ignore
+    /// # let r = Ok::<_, ()>(());
+    /// assert!(r.is_ok());
+    /// ```
+    #[clippy::version = "1.64.0"]
+    pub ASSERT_OK,
+    style,
+    "`assert!(r.is_ok())` gives worse error message than directly calling `r.unwrap()`"
+}
+
+declare_lint_pass!(AssertOk => [ASSERT_OK]);
+
+impl<'tcx> LateLintPass<'tcx> for AssertOk {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) {
+        if_chain! {
```

Comment:
This should use let chains instead of `if_chain!`. i.e.
```rust
if let pat = expr
    && let pat = expr
{}
```

Reply:
Is there a particular reason that let chain is better? I see a few `let`s with dependency which can't be collapsed, and a few non-`let` conditions as well.